spec/models/manageiq/providers/autosde/inventory_spec.rb

```diff
@@ -5,23 +5,35 @@
                       :hostname => Rails.application.secrets.autosde[:appliance_host])
   end
 
+  let(:first_wwpn) { "2100000E1EE89D90" }
+
   it "gets inventory from the appliance" do
     VCR.use_cassette("ems_refresh") do
       EmsRefresh.refresh(ems)
       ems.reload
     end
 
     # expected that refresh will bring all components
-    expect(ems.physical_storages).to_not be_empty
-    expect(ems.physical_storage_families).to_not be_empty
-    expect(ems.physical_storages.first.physical_storage_family).to eq(ems.physical_storage_families.find_by(:name=>'svc'))
+    expect(ems.physical_storages).to_not(be_empty)
+    expect(ems.physical_storage_families).to_not(be_empty)
+    expect(ems.physical_storages.first.physical_storage_family).to(eq(ems.physical_storage_families.find_by(:name=>'svc')))
+
+    expect(ems.storage_resources).to_not(be_empty)
+    expect(ems.storage_resources.first.physical_storage).to(eq(ems.physical_storages.first))
 
-    expect(ems.storage_resources).to_not be_empty
-    expect(ems.storage_resources.first.physical_storage).to eq(ems.physical_storages.first)
+    expect(ems.storage_services).to_not(be_empty)
+    expect(ems.cloud_volumes).to_not(be_empty)
 
-    expect(ems.storage_services).to_not be_empty
-    expect(ems.cloud_volumes).to_not be_empty
+    expect(ems.cloud_volumes.first).to(be_instance_of(ManageIQ::Providers::Autosde::StorageManager::CloudVolume))
 
-    expect(ems.cloud_volumes.first).to be_instance_of(ManageIQ::Providers::Autosde::StorageManager::CloudVolume)
+    # wwpn candidates
+    expect(ems.wwpn_candidates.count).to(eq(2))
+    first_candidate = ems.wwpn_candidates.first
+    expect(first_candidate).to(have_attributes(
+                                 :candidate        => first_wwpn,
+                                 :ems_id           => ems.id,
+                                 :physical_storage => ems.physical_storages.find_by(:ems_id => ems.id)
```

Comment:
`ems.physical_storage.find_by(:ems_id => ems.id)` is redudant (ems.physical_storages already does that), think you meant to do `.find_by(:ems_ref => "ref-of-your-physical-storage")`

Reply:
For wwpn_candidates, we never going to create them, only read them
But persister need some uniq key, when it builds InventoryObject in InventoryCollection. I needed workaround.
I've recalled I was using ems_ref as such key, and gave it the candidate value (it is uniq wordlwide).
So in the test I need only to check if wwpn_candidate is created as expected, and not to use ems_ref for finding anything
Thus, knowing ems or physical_storage,
I can get wwpm_candidates like:
* ems.wwpn_candidates
* ems.physical_storages.first.wwpn_candidates


From app/models/manageiq/providers/autosde/inventory/parser/storage_manager.rb:

`def wwpn_candidates
    collector.wwpn_candidates.each do |candidate|
      persister.wwpn_candidates.build(
        :candidate        => candidate.wwpn,
        :ems_ref          => candidate.wwpn,
        :physical_storage => persister.physical_storages.lazy_find(candidate.system_uuid)
      )
    end
  end`

Reply:
Sorry not sure what you mean.  You're doing `ems.physical_storages.find_by(:ems_id => ems.id)` which doesn't make any sense.  If you have multiple physical storages and you need to find one (and find the same one consistently) then you need to `.find_by(:ems_ref => "ref-of-the physical-storage")`

Reply:
High level, this spec is just querying the database.  It has nothing to do with creating or reading wwpn_candidates, all of that has already happened in the refresher.

Now you just need to query the DB to find the same records consistently so that your specs don't fail randomly.  That's the problem with using `first_candidate = ems.wwpn_candidates.first`